src/Chainweb/Pact/PactService.hs

```diff
@@ -1269,66 +1301,75 @@ rewindTo
     :: forall cas . PayloadCasLookup cas
     => Maybe BlockHeight
         -- ^ if set, limit rewinds to this delta
-    -> Maybe (BlockHeight, ParentHash)
-        -- ^ The block height @height@ to which to restore and the parent header
-        -- @parent@.
-        --
-        -- It holds that @(_blockHeight parent == pred height)@
+    -> Maybe ParentHeader
+        -- ^ The parent header which is the rewind target
     -> PactServiceM cas ()
-rewindTo rewindLimit = maybe rewindGenesis doRewind
+rewindTo _ Nothing = return ()
+rewindTo rewindLimit (Just (ParentHeader parent)) = do
+
+    -- skip if the checkpointer is already at the target.
+    (_, lastHash) <- getCheckpointer >>= liftIO . _cpGetLatestBlock >>= \case
+        Nothing -> throwM NoBlockValidatedYet
+        Just p -> return p
+
+    if lastHash == parentHash
+      then setParentHeader "rewindTo" (ParentHeader parent)
```

Comment:
Might be good to note where `setParentHeader` gets called in the `else` case (presumably in `playFork -> withCheckpointerWithoutRewind` right?)

Reply:
hmm, maybe it shouldn't be set here. `withCheckpointerWithoutRewind` sets it only if the result is saved. One could argue that here it doesn't matter, because in the case of `lastHash == parentHash` it doesn't make a difference whether the result is saved or discarded. But, under that assumption, setting '_psParentHash' here should not be needed in first place.

Reply:
The current approach is a bit paranoid. It makes sure that it is set before validation starts and also after it ends. In theory the latter should be sufficient (assuming proper initialization). The former is added as a second line of defense in case of failures slipping through existing finalizers.